ccl/importccl: TestImportIntoCSV failed

[(ccl/importccl).TestImportIntoCSV failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2076322&tab=buildLog) on [master@1b5d070c93375d3e14c146241e8bafde349529bd](https://github.com/cockroachdb/cockroach/commits/1b5d070c93375d3e14c146241e8bafde349529bd):

```
=== RUN   TestImportIntoCSV/import-into-with-opts


ERROR: exit status 2

35 runs completed, 1 failures, over 11m40s
context canceled

```

<details><summary>More</summary><p>
Parameters:

- TAGS=
- GOFLAGS=-parallel=4

```
make stressrace TESTS=TestImportIntoCSV PKG=./pkg/ccl/importccl TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestImportIntoCSV.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(ccl/importccl).TestImportIntoCSV failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2270704&tab=buildLog) on [master@897b15423eb7b0ab2706c493991a1681dce72351](https://github.com/cockroachdb/cockroach/commits/897b15423eb7b0ab2706c493991a1681dce72351):

```
=== RUN   TestImportIntoCSV
--- FAIL: TestImportIntoCSV (151.39s)
    test_log_scope.go:154: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestImportIntoCSV314129020
    test_log_scope.go:63: use -show-logs to present logs inline
=== RUN   TestImportIntoCSV/target-cols-with-default-pk/#00
        --- FAIL: TestImportIntoCSV/target-cols-with-default-pk/#00 (2.26s)
            import_stmt_test.go:2665: error executing 'IMPORT INTO t (a) CSV DATA ($1)': pq: another operation is currently operating on the table
=== RUN   TestImportIntoCSV/target-cols-with-default-pk
    --- FAIL: TestImportIntoCSV/target-cols-with-default-pk (10.15s)

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestImportIntoCSV PKG=./pkg/ccl/importccl TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestImportIntoCSV.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(ccl/importccl).TestImportIntoCSV failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2273852&tab=buildLog) on [master@02aa876f357692e6ab71967289a05ffc2208de07](https://github.com/cockroachdb/cockroach/commits/02aa876f357692e6ab71967289a05ffc2208de07):

```
                			d INT8,
                			e INT8,
                			f STRING)': pq: relation "t" already exists
=== RUN   TestImportIntoCSV/csv-with-more-than-targeted-columns
    --- FAIL: TestImportIntoCSV/csv-with-more-than-targeted-columns (0.03s)
        import_stmt_test.go:2733: error executing 'CREATE TABLE t (a INT PRIMARY KEY, b STRING)': pq: relation "t" already exists
=== RUN   TestImportIntoCSV/more-table-cols-than-csv
    --- FAIL: TestImportIntoCSV/more-table-cols-than-csv (0.02s)
        import_stmt_test.go:2762: error executing 'CREATE TABLE t (a INT, b STRING, c INT)': pq: relation "t" already exists
=== RUN   TestImportIntoCSV/no-target-cols-specified
    --- FAIL: TestImportIntoCSV/no-target-cols-specified (0.02s)
        import_stmt_test.go:2833: error executing 'CREATE TABLE t (a INT PRIMARY KEY, b STRING)': pq: relation "t" already exists
=== RUN   TestImportIntoCSV/import-into-no-decompress-gzip#01
    --- FAIL: TestImportIntoCSV/import-into-no-decompress-gzip#01 (0.02s)
        import_stmt_test.go:2475: error executing 'CREATE TABLE t (a INT, b STRING)': pq: relation "t" already exists
=== RUN   TestImportIntoCSV/target-cols-with-explicit-pk
    --- FAIL: TestImportIntoCSV/target-cols-with-explicit-pk (0.03s)
        import_stmt_test.go:2694: error executing 'CREATE TABLE t (a INT PRIMARY KEY, b STRING)': pq: relation "t" already exists
=== RUN   TestImportIntoCSV/target-cols-excluding-explicit-pk
    --- FAIL: TestImportIntoCSV/target-cols-excluding-explicit-pk (0.02s)
        import_stmt_test.go:2748: error executing 'CREATE TABLE t (a INT PRIMARY KEY, b STRING)': pq: relation "t" already exists
=== RUN   TestImportIntoCSV/import-into-unique_rowid
    --- FAIL: TestImportIntoCSV/import-into-unique_rowid (0.03s)
        import_stmt_test.go:2544: error executing 'CREATE TABLE t (a INT, b STRING)': pq: relation "t" already exists
=== RUN   TestImportIntoCSV/import-into-checkpoint-leftover
    --- FAIL: TestImportIntoCSV/import-into-checkpoint-leftover (0.03s)
        import_stmt_test.go:2571: error executing 'CREATE TABLE t (a INT PRIMARY KEY, b STRING)': pq: relation "t" already exists
=== RUN   TestImportIntoCSV/offline-state
    --- FAIL: TestImportIntoCSV/offline-state (0.02s)
        import_stmt_test.go:2595: error executing 'CREATE TABLE t (a INT PRIMARY KEY, b STRING)': pq: relation "t" already exists
=== RUN   TestImportIntoCSV/target-cols-with-default-pk/#02
        --- FAIL: TestImportIntoCSV/target-cols-with-default-pk/#02 (0.02s)
            import_stmt_test.go:2681: error executing 'CREATE TABLE t (a INT8,
                			b INT8,
                			c STRING,
                			d INT8,
                			e INT8,
                			f STRING)': pq: relation "t" already exists
=== RUN   TestImportIntoCSV/multiple-import-into-without-pk
    --- FAIL: TestImportIntoCSV/multiple-import-into-without-pk (0.02s)
        import_stmt_test.go:2900: error executing 'CREATE TABLE t (a INT, b STRING)': pq: relation "t" already exists
=== RUN   TestImportIntoCSV/import-into-dups-in-sst
    --- FAIL: TestImportIntoCSV/import-into-dups-in-sst (0.03s)
        import_stmt_test.go:2974: error executing 'CREATE TABLE t (a INT PRIMARY KEY, b STRING)': pq: relation "t" already exists
=== RUN   TestImportIntoCSV/fewer-table-cols-than-csv
    --- FAIL: TestImportIntoCSV/fewer-table-cols-than-csv (0.02s)
        import_stmt_test.go:2821: error executing 'CREATE TABLE t (a INT)': pq: relation "t" already exists
=== RUN   TestImportIntoCSV/import-into-key-collision
    --- FAIL: TestImportIntoCSV/import-into-key-collision (0.03s)
        import_stmt_test.go:2992: error executing 'CREATE TABLE t (a INT PRIMARY KEY, b STRING)': pq: relation "t" already exists

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestImportIntoCSV PKG=./pkg/ccl/importccl TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestImportIntoCSV.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(ccl/importccl).TestImportIntoCSV failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2279382&tab=buildLog) on [master@1c2b2729a7475c9357e048920c5e98cb6b155678](https://github.com/cockroachdb/cockroach/commits/1c2b2729a7475c9357e048920c5e98cb6b155678):

```
=== RUN   TestImportIntoCSV
--- FAIL: TestImportIntoCSV (45.29s)
    test_log_scope.go:154: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestImportIntoCSV258702598
    test_log_scope.go:63: use -show-logs to present logs inline
=== RUN   TestImportIntoCSV/import-into-invalidate-constraints
    --- FAIL: TestImportIntoCSV/import-into-invalidate-constraints (0.48s)
        import_stmt_test.go:3010: error executing 'CREATE TABLE t (a INT CHECK (a >= 0), b STRING, CONSTRAINT fk_ref FOREIGN KEY (b) REFERENCES ref)': pq: relation "t" already exists
=== RUN   TestImportIntoCSV/import-into-key-collision
    --- FAIL: TestImportIntoCSV/import-into-key-collision (1.01s)
        import_stmt_test.go:2999: expected error 'ingested key collides with an existing one: /Table/\d+/1/0/0', got: pq: job 590780311106486273: cannot be reverted, manual cleanup may be required: rolling back partially completed IMPORT: only offline tables can be reverted
        panic.go:563: error executing 'DROP TABLE t': pq: relation "t" does not exist

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestImportIntoCSV PKG=./pkg/ccl/importccl TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestImportIntoCSV.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(ccl/importccl).TestImportIntoCSV failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2285166&tab=buildLog) on [master@3202b174b201e8db099578b93d6dd9396ccf874b](https://github.com/cockroachdb/cockroach/commits/3202b174b201e8db099578b93d6dd9396ccf874b):

```
    --- FAIL: TestImportIntoCSV/target-cols-with-default-pk (0.10s)
=== RUN   TestImportIntoCSV/import-not-targeted-not-null
    --- FAIL: TestImportIntoCSV/import-not-targeted-not-null (0.04s)
        import_stmt_test.go:2860: error executing 'CREATE TABLE t (a INT, b INT NOT NULL)': pq: relation "t" already exists
=== RUN   TestImportIntoCSV/import-into-no-glob-wildcard
    --- FAIL: TestImportIntoCSV/import-into-no-glob-wildcard (0.04s)
        import_stmt_test.go:2475: error executing 'CREATE TABLE t (a INT, b STRING)': pq: relation "t" already exists
=== RUN   TestImportIntoCSV/import-into-unique_rowid
    --- FAIL: TestImportIntoCSV/import-into-unique_rowid (0.03s)
        import_stmt_test.go:2544: error executing 'CREATE TABLE t (a INT, b STRING)': pq: relation "t" already exists
=== RUN   TestImportIntoCSV/target-cols-with-default-pk/#02
        --- FAIL: TestImportIntoCSV/target-cols-with-default-pk/#02 (0.03s)
            import_stmt_test.go:2681: error executing 'CREATE TABLE t (a INT8,
                			b INT8,
                			c STRING,
                			d INT8,
                			e INT8,
                			f STRING)': pq: relation "t" already exists
=== RUN   TestImportIntoCSV/target-cols-excluding-explicit-pk
    --- FAIL: TestImportIntoCSV/target-cols-excluding-explicit-pk (0.03s)
        import_stmt_test.go:2748: error executing 'CREATE TABLE t (a INT PRIMARY KEY, b STRING)': pq: relation "t" already exists
=== RUN   TestImportIntoCSV/no-target-cols-specified
    --- FAIL: TestImportIntoCSV/no-target-cols-specified (0.04s)
        import_stmt_test.go:2833: error executing 'CREATE TABLE t (a INT PRIMARY KEY, b STRING)': pq: relation "t" already exists
=== RUN   TestImportIntoCSV/multiple-import-into-without-pk
    --- FAIL: TestImportIntoCSV/multiple-import-into-without-pk (0.03s)
        import_stmt_test.go:2900: error executing 'CREATE TABLE t (a INT, b STRING)': pq: relation "t" already exists
=== RUN   TestImportIntoCSV/multiple-file-import-into-without-pk
    --- FAIL: TestImportIntoCSV/multiple-file-import-into-without-pk (0.03s)
        import_stmt_test.go:2929: error executing 'CREATE TABLE t (a INT, b STRING)': pq: relation "t" already exists
=== RUN   TestImportIntoCSV/import-into-same-file-diff-imports
    --- FAIL: TestImportIntoCSV/import-into-same-file-diff-imports (0.03s)
        import_stmt_test.go:2951: error executing 'CREATE TABLE t (a INT PRIMARY KEY, b STRING)': pq: relation "t" already exists
=== RUN   TestImportIntoCSV/import-into-checkpoint-leftover
    --- FAIL: TestImportIntoCSV/import-into-checkpoint-leftover (0.03s)
        import_stmt_test.go:2571: error executing 'CREATE TABLE t (a INT PRIMARY KEY, b STRING)': pq: relation "t" already exists
=== RUN   TestImportIntoCSV/target-cols-with-default-pk/#00
        --- FAIL: TestImportIntoCSV/target-cols-with-default-pk/#00 (0.03s)
            import_stmt_test.go:2661: error executing 'CREATE TABLE t (a INT8,
                			b INT8,
                			c STRING,
                			d INT8,
                			e INT8,
                			f STRING)': pq: relation "t" already exists
=== RUN   TestImportIntoCSV/target-cols-with-explicit-pk
    --- FAIL: TestImportIntoCSV/target-cols-with-explicit-pk (0.03s)
        import_stmt_test.go:2694: error executing 'CREATE TABLE t (a INT PRIMARY KEY, b STRING)': pq: relation "t" already exists
=== RUN   TestImportIntoCSV/import-into-invalidate-constraints
    --- FAIL: TestImportIntoCSV/import-into-invalidate-constraints (1.61s)
        import_stmt_test.go:3010: error executing 'CREATE TABLE t (a INT CHECK (a >= 0), b STRING, CONSTRAINT fk_ref FOREIGN KEY (b) REFERENCES ref)': pq: relation "t" already exists

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestImportIntoCSV PKG=./pkg/ccl/importccl TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestImportIntoCSV.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(ccl/importccl).TestImportIntoCSV failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2285309&tab=buildLog) on [master@a6dc5e931154c2e7fff1f00ac4d6b272978be990](https://github.com/cockroachdb/cockroach/commits/a6dc5e931154c2e7fff1f00ac4d6b272978be990):

```
=== RUN   TestImportIntoCSV/target-cols-with-explicit-pk
    --- FAIL: TestImportIntoCSV/target-cols-with-explicit-pk (0.05s)
        import_stmt_test.go:2694: error executing 'CREATE TABLE t (a INT PRIMARY KEY, b STRING)': pq: relation "t" already exists
=== RUN   TestImportIntoCSV/csv-with-more-than-targeted-columns
    --- FAIL: TestImportIntoCSV/csv-with-more-than-targeted-columns (0.05s)
        import_stmt_test.go:2733: error executing 'CREATE TABLE t (a INT PRIMARY KEY, b STRING)': pq: relation "t" already exists
=== RUN   TestImportIntoCSV/cols-named-with-reserved-keywords
    --- FAIL: TestImportIntoCSV/cols-named-with-reserved-keywords (0.04s)
        import_stmt_test.go:2802: error executing 'CREATE TABLE t ("select" INT PRIMARY KEY, "from" INT, "Some-c,ol-'Name'" STRING NOT NULL)': pq: relation "t" already exists
=== RUN   TestImportIntoCSV/multiple-import-into-without-pk
    --- FAIL: TestImportIntoCSV/multiple-import-into-without-pk (0.08s)
        import_stmt_test.go:2900: error executing 'CREATE TABLE t (a INT, b STRING)': pq: relation "t" already exists
=== RUN   TestImportIntoCSV/import-into-same-file-diff-imports
    --- FAIL: TestImportIntoCSV/import-into-same-file-diff-imports (0.05s)
        import_stmt_test.go:2951: error executing 'CREATE TABLE t (a INT PRIMARY KEY, b STRING)': pq: relation "t" already exists
=== RUN   TestImportIntoCSV/import-into-dups-in-sst
    --- FAIL: TestImportIntoCSV/import-into-dups-in-sst (0.05s)
        import_stmt_test.go:2974: error executing 'CREATE TABLE t (a INT PRIMARY KEY, b STRING)': pq: relation "t" already exists
=== RUN   TestImportIntoCSV/offline-state
    --- FAIL: TestImportIntoCSV/offline-state (0.05s)
        import_stmt_test.go:2595: error executing 'CREATE TABLE t (a INT PRIMARY KEY, b STRING)': pq: relation "t" already exists
=== RUN   TestImportIntoCSV/target-cols-with-default-pk/#01
        --- FAIL: TestImportIntoCSV/target-cols-with-default-pk/#01 (0.05s)
            import_stmt_test.go:2671: error executing 'CREATE TABLE t (a INT8,
                			b INT8,
                			c STRING,
                			d INT8,
                			e INT8,
                			f STRING)': pq: relation "t" already exists
=== RUN   TestImportIntoCSV/target-cols-with-default-pk
    --- FAIL: TestImportIntoCSV/target-cols-with-default-pk (0.14s)
=== RUN   TestImportIntoCSV/fewer-table-cols-than-csv
    --- FAIL: TestImportIntoCSV/fewer-table-cols-than-csv (0.05s)
        import_stmt_test.go:2821: error executing 'CREATE TABLE t (a INT)': pq: relation "t" already exists
=== RUN   TestImportIntoCSV/no-target-cols-specified
    --- FAIL: TestImportIntoCSV/no-target-cols-specified (0.04s)
        import_stmt_test.go:2833: error executing 'CREATE TABLE t (a INT PRIMARY KEY, b STRING)': pq: relation "t" already exists
=== RUN   TestImportIntoCSV/multiple-file-import-into-without-pk
    --- FAIL: TestImportIntoCSV/multiple-file-import-into-without-pk (0.06s)
        import_stmt_test.go:2929: error executing 'CREATE TABLE t (a INT, b STRING)': pq: relation "t" already exists
=== RUN   TestImportIntoCSV/import-into-no-glob-wildcard
    --- FAIL: TestImportIntoCSV/import-into-no-glob-wildcard (2.02s)
        import_stmt_test.go:2495: expected error 'pq: (.+) no such file or directory', got: pq: job 591888501067612161: cannot be reverted, manual cleanup may be required: rolling back partially completed IMPORT: only offline tables can be reverted
        panic.go:563: error executing 'DROP TABLE t': pq: relation "t" does not exist
=== RUN   TestImportIntoCSV/import-into-checkpoint-leftover
    --- FAIL: TestImportIntoCSV/import-into-checkpoint-leftover (0.05s)
        import_stmt_test.go:2571: error executing 'CREATE TABLE t (a INT PRIMARY KEY, b STRING)': pq: relation "t" already exists
=== RUN   TestImportIntoCSV/more-table-cols-than-csv
    --- FAIL: TestImportIntoCSV/more-table-cols-than-csv (0.04s)
        import_stmt_test.go:2762: error executing 'CREATE TABLE t (a INT, b STRING, c INT)': pq: relation "t" already exists

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestImportIntoCSV PKG=./pkg/ccl/importccl TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestImportIntoCSV.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/sql/catalog/descs/collection_test.go
+++ b/pkg/sql/catalog/descs/collection_test.go
@@ -16,11 +16,14 @@ import (
 
 	"github.com/cockroachdb/cockroach/pkg/base"
 	"github.com/cockroachdb/cockroach/pkg/kv"
+	"github.com/cockroachdb/cockroach/pkg/kv/kvserver"
+	"github.com/cockroachdb/cockroach/pkg/roachpb"
 	"github.com/cockroachdb/cockroach/pkg/sql/catalog/descpb"
 	"github.com/cockroachdb/cockroach/pkg/sql/catalog/descs"
 	"github.com/cockroachdb/cockroach/pkg/sql/catalog/lease"
 	"github.com/cockroachdb/cockroach/pkg/sql/catalog/tabledesc"
 	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
+	"github.com/cockroachdb/cockroach/pkg/sql/sqlutil"
 	"github.com/cockroachdb/cockroach/pkg/testutils/sqlutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/testcluster"
 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
@@ -90,4 +93,68 @@ func TestCollectionWriteDescToBatch(t *testing.T) {
 		require.Equal(t, newTable, newTableResolved)
 		return txn.Run(ctx, b)
 	}))
+}
+
+// TestTxnClearsCollectionOnRetry tests that descs.Txn() correctly clears the
+// state of its associated descs.Collection on a retry error at commit time.
+// Regression test for #51197.
+func TestTxnClearsCollectionOnRetry(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	ctx := context.Background()
+
+	const txnName = "descriptor update"
+	haveInjectedRetry := false
+	var serverArgs base.TestServerArgs
+	params := base.TestClusterArgs{ServerArgs: serverArgs}
+	params.ServerArgs.Knobs.Store = &kvserver.StoreTestingKnobs{
+		TestingRequestFilter: func(ctx context.Context, r roachpb.BatchRequest) *roachpb.Error {
+			if r.Txn == nil || r.Txn.Name != txnName {
+				return nil
+			}
+			if _, ok := r.GetArg(roachpb.EndTxn); ok {
+				if !haveInjectedRetry {
+					haveInjectedRetry = true
+					// Force a retry error the first time.
+					return roachpb.NewError(roachpb.NewTransactionRetryError(roachpb.RETRY_REASON_UNKNOWN, "injected error"))
+				}
+			}
+			return nil
+		},
+	}
+	tc := testcluster.StartTestCluster(t, 1, params)
+	defer tc.Stopper().Stop(ctx)
+
+	s := tc.Server(0)
+
+	tdb := sqlutils.MakeSQLRunner(tc.ServerConn(0))
+	tdb.Exec(t, `CREATE DATABASE db`)
+	tdb.Exec(t, `USE db`)
+	tdb.Exec(t, `CREATE SCHEMA schema`)
+	tdb.Exec(t, `CREATE TABLE db.schema.table()`)
+	tn := tree.MakeTableNameWithSchema("db", "schema", "table")
+
+	err := descs.Txn(
+		ctx,
+		s.ClusterSettings(),
+		s.LeaseManager().(*lease.Manager),
+		s.InternalExecutor().(sqlutil.InternalExecutor),
+		s.DB(),
+		func(ctx context.Context, txn *kv.Txn, descriptors *descs.Collection) error {
+			txn.SetDebugName(txnName)
+
+			flags := tree.ObjectLookupFlagsWithRequired()
+			flags.RequireMutable = true
+			desc, err := descriptors.GetMutableTableDescriptor(ctx, txn, &tn, flags)
+			require.NoError(t, err)
+			// Verify that the descriptor version is always 1 prior to the write and 2
+			// after the write even after a retry.
+			require.Equal(t, descpb.DescriptorVersion(1), desc.Version)
+			require.NoError(t, descriptors.WriteDesc(ctx, false /* kvTrace */, desc, txn))
+			require.Equal(t, descpb.DescriptorVersion(2), desc.Version)
+			return nil
+		},
+	)
+	require.NoError(t, err)
 }

--- a/pkg/sql/catalog/descs/txn.go
+++ b/pkg/sql/catalog/descs/txn.go
@@ -51,14 +51,13 @@ func Txn(
 	if descsCol.databaseLeasingUnsupported {
 		descsCol.databaseCache = database.NewCache(leaseMgr.Codec(), &config.SystemConfig{})
 	}
-	defer descsCol.ReleaseAll(ctx)
 	for {
 		if err := db.Txn(ctx, func(ctx context.Context, txn *kv.Txn) error {
+			defer descsCol.ReleaseAll(ctx)
 			if err := txn.SetSystemConfigTrigger(leaseMgr.Codec().ForSystemTenant()); err != nil {
 				return err
 			}
 			if err := f(ctx, txn, descsCol); err != nil {
-				descsCol.ReleaseAll(ctx)
 				return err
 			}
 			retryErr, err := CheckTwoVersionInvariant(
@@ -68,7 +67,7 @@ func Txn(
 			}
 			return err
 		}); errors.Is(err, errTwoVersionInvariantViolated) {
-			descsCol.ReleaseAll(ctx)
+			continue
 		} else {
 			return err
 		}

